plugins/basic/nexus-repository-docker/src/main/java/org/sonatype/nexus/repository/docker/internal/DockerFacetImpl.java

```diff
@@ -0,0 +1,669 @@
+
+      indexStream = streamSupplier.get();
+    }
+    else {
+      // merge existing index with the new one
+      List<Map<String, Object>> existingIndex = readImagesIndex(blobStreamSupplier(tx, imagesIndexAsset));
+      List<Map<String, Object>> newIndex = readImagesIndex(streamSupplier);
+      List<Map<String, Object>> mergedIndex = new ArrayList<>(existingIndex);
+      for (Map<String, Object> newImage : newIndex) {
+        boolean found = false;
+        for (Map<String, Object> existingImage : existingIndex) {
+          if (Objects.equals(existingImage.get("id"), newImage.get("id")) &&
+              Objects.equals(existingImage.get("Tag"), newImage.get("Tag"))) {
+            existingImage.putAll(newImage);
+            found = true;
+            break;
+          }
+        }
+        if (!found) {
+          mergedIndex.add(newImage);
+        }
+      }
+      indexStream = new ByteArrayInputStream(mapper.writeValueAsBytes(mergedIndex));
+    }
+
+    AttributesMap contentAttributes = null;
+    if (payload instanceof Content) {
+      contentAttributes = ((Content) payload).getAttributes();
+    }
+    Content.applyToAsset(imagesIndexAsset, Content.maintainLastModified(imagesIndexAsset, contentAttributes));
+    tx.setBlob(imagesIndexAsset, imagesIndexAssetName, indexStream, hashAlgorithms, null, payload.getContentType());
+    tx.saveAsset(imagesIndexAsset);
+  }
+
+  @Override
+  @Transactional(retryOn = IllegalStateException.class)
+  public Payload getLayerMetadata(final String layerId) throws IOException {
+    validateLayerId(layerId);
+    StorageTx tx = UnitOfWork.currentTransaction();
+
+    String metadataAssetName = layerId + "/json";
+
+    Bucket bucket = tx.getBucket();
+    Asset metadataAsset = findAsset(tx, bucket, metadataAssetName);
+    // check that we have the layer and layer content was fully uploaded
+    if (metadataAsset == null || Boolean.TRUE.equals(metadataAsset.formatAttributes().get(P_IN_PROGRESS))) {
+      throw new V1Exception("Image not found", HttpStatus.NOT_FOUND);
+    }
+
+    Content content = marshall(metadataAsset, tx.requireBlob(metadataAsset.requireBlobRef()));
+
+    NestedAttributesMap componentAttributes = metadataAsset.formatAttributes();
+    content.getAttributes().set(P_X_DOCKER_SIZE, componentAttributes.get(P_X_DOCKER_SIZE));
+    content.getAttributes().set(P_X_DOCKER_CHECKSUM_PAYLOAD, componentAttributes.get(P_X_DOCKER_CHECKSUM_PAYLOAD));
+
+    return content;
+  }
+
+  @Override
+  public void putLayerMetadata(final String layerId, final Payload content) throws IOException {
+    validateLayerId(layerId);
+    try (TempStreamSupplier streamSupplier = new TempStreamSupplier(content.openInputStream())) {
+      doPutLayerMetadata(layerId, streamSupplier, content);
+    }
+  }
+
+  @Transactional(retryOn = {ONeedRetryException.class, ORecordDuplicatedException.class})
+  protected void doPutLayerMetadata(final String layerId,
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "doPutLayerMetadata" is 15 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)